### PR TITLE
feat(db): use a materialized view for ledger updates

### LINKED ISSUE #686: Turn `stardust_ledger_updates` into an on-demand materialized view

### PR DESCRIPTION
**Pull request by Alexandcoats**
Closes #686

### CODE REVIEW

File: src/db/collections/outputs/mod.rs [hunk 1]
```diff
@@ -213,6 +214,68 @@ impl OutputCollection {
...
+                    "_id": {
+                        "milestone_index": { "$ifNull": [
+                            "$metadata.spent_metadata.spent.milestone_index",
+                            "$metadata.booked.milestone_index"
+                        ] },
+                        "output_id": "$_id",
+                        "is_spent": { "$ne": [ "$metadata.spent_metadata", null ] },
+                    },
+                    "address": "$details.address",
+                    "milestone_timestamp": "$metadata.booked.milestone_timestamp",
```
Comment (Alex6323) [inline src/db/collections/outputs/mod.rs:267]: Does that always have to be the `timestamp` of the "booking" milestone?

Comment (Alexandcoats) [reply src/db/collections/outputs/mod.rs:267]: Ah, good catch.

Comment (grtlr) [reply src/db/collections/outputs/mod.rs]: Good catch, I think you're right. This should probably use the `$ifNull` mechanism too.

@Alexandcoats Could you please at this to your test case as well?

File: src/db/collections/outputs/mod.rs [hunk 2]
```diff
@@ -213,6 +214,74 @@ impl OutputCollection {
...
+                        ] },
+                        "output_id": "$_id",
+                        "is_spent": { "$ne": [ "$metadata.spent_metadata", null ] },
+                    },
+                    "address": "$details.address",
+                    "milestone_timestamp": { "$ifNull": [
+                        "$metadata.spent_metadata.spent.milestone_timestamp",
+                        "$metadata.booked.milestone_timestamp"
+                    ] },
+                } },
```
Comment (grtlr) [inline src/db/collections/outputs/mod.rs:237]: As the `$project` stage is identical I would love to pull this out into a function to reduce the redundancy.

Comment (Alexandcoats) [reply src/db/collections/outputs/mod.rs:237]: Let's worry about this as part of our refactoring of these later.

File: src/bin/inx-chronicle/stardust_inx/mod.rs [hunk 3]
```diff
@@ -187,6 +185,8 @@ impl Actor for InxWorker {
                 task.await.unwrap()?;
             }
 
+            self.db.collection::<OutputCollection>().create_ledger_updates().await?;
```
Comment (grtlr) [inline src/bin/inx-chronicle/stardust_inx/mod.rs:188]:
```suggestion
            self.db.collection::<OutputCollection>().create_ledger_updates_view().await?;
```

File: src/bin/inx-chronicle/stardust_inx/mod.rs [hunk 4]
```diff
@@ -256,6 +256,10 @@ impl HandleEvent<Result<LedgerUpdateRecord, InxError>> for InxWorker {
 
         insert_unspent_outputs(&self.db, ledger_update.created).await?;
         update_spent_outputs(&self.db, ledger_update.consumed).await?;
+        self.db
+            .collection::<OutputCollection>()
+            .merge_into_ledger_updates(ledger_update.milestone_index)
```
Comment (grtlr) [inline src/bin/inx-chronicle/stardust_inx/mod.rs:261]:
```suggestion
            .merge_into_ledger_updates_view(ledger_update.milestone_index)
```

File: src/db/collections/outputs/mod.rs (hunk 2 above)
Comment (Alexandcoats) [reply src/db/collections/outputs/mod.rs:237]: I mean pulling out the pipeline stages into functions or whatever.

Comment (grtlr) [reply src/db/collections/outputs/mod.rs:237]: We should grab these low-hanging fruits though, especially if we have already identified areas for improvement.

Comment (Alexandcoats) [reply src/db/collections/outputs/mod.rs:237]: Unfortunately it is not low hanging in my opinion. The right solution is not as clear as simply pulling this into a function. I think we need to give it some thought and apply a uniform solution everywhere.

**Review by grtlr — APPROVED** [review]

**Pull request merged**